File: 3-sum-of-digits-to-power/C/digitsum.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <limits.h>

unsigned long cache[10];

unsigned long max_num(int exponent) {
  int digit = 2;
  unsigned long nine_to_the_exponent = pow(9, exponent);
  unsigned long max_digit_sum = nine_to_the_exponent * digit;
  unsigned long maximum = pow(10, digit);
  while (maximum - 1 < max_digit_sum) {
    maximum *= 10;
    digit++;
    max_digit_sum = nine_to_the_exponent * digit;
  }
  return max_digit_sum;
}

unsigned long sum_digits(unsigned long num, int exponent) {
  unsigned long sum = 0;
  for (; num != 0; num /= 10) {
    sum += cache[num % 10];
  }
  return sum;
}

void matching_numbers(int exponent) {
  unsigned long max = max_num(exponent);
  unsigned long step = 10;
  for (int j = 0; j < step; j++) {
    cache[j] = pow(j, exponent);
  }
  
  // find matching numbers from 10..max
  for (unsigned long i = 10; i < (max + step); i += step) {
    unsigned long base_sum = sum_digits(i, exponent);
    for (unsigned long j = 0; j < step; j++) {
      if (j + i == (base_sum + cache[j])) {
        printf("%ld, ", j + i);
      }
    }
  }
  printf("\n");
}

int parse_args(int argc, char **argv) {
  int retval = 5;
  if (argc > 1) {
    retval = atoi(argv[1]);
  }
  return retval;
}

int main(int argc, char **argv) {
  int exponent = parse_args(argc, argv);
  printf("%d: ", exponent);
  matching_numbers(exponent);
  return 0;
}

File: 3-sum-of-digits-to-power/go/digitsum.go
package main

import (
	"flag"
	"fmt"
	"math"
	"sync"
)

var cache [10]int64

func max_num(exponent float64) int64 {
	var max_digit_sum int64 = 9 + 9
	var maximum int64 = 10
	nine_to_the_exponent := int64(math.Floor(math.Pow(9, exponent)))
	var digit int64 = 2
	for ; maximum-1 < max_digit_sum; digit++ {
		maximum *= 10
		max_digit_sum = nine_to_the_exponent * digit
	}
	return max_digit_sum
}

func sum_digits(num int64) int64 {
	var sum_of_digits_to_power int64 = 0
	for remainder := num; remainder > 0; remainder /= 10 {
		sum_of_digits_to_power += cache[remainder%10]
	}
	return sum_of_digits_to_power
}

func find_matches_in_next_10(base_sum int64, base_num int64) []int64 {
	var nums []int64
	var i int64 = 0
	for ; i < 10; i++ {
		if (base_num + i) == (base_sum + cache[i]) {
			nums = append(nums, (i + base_num))
		}
	}
	return nums
}

func matching_numbers_between(start int64, end int64, _ float64, wg *sync.WaitGroup) {
	for i := start; i < end; i += int64(10) {
		base_sum := sum_digits(i)
		nums := find_matches_in_next_10(base_sum, int64(i))
		for _, n := range nums {
			fmt.Printf("%d, ", n)
		}
	}
	wg.Done()
}

func matching_numbers(exponent float64) {
	max := max_num(exponent)
	var wg = &sync.WaitGroup{}
	var start int64 = 10
	fmt.Printf("start: %d .. max: %d: ", start, max)

	// find matching numbers from 10..max
	// assign each working group a disjoint set of consecutive numbers to search
	var i int64 = start
	const increment int64 = 1_000_000
	for ; i < max+increment; i += increment {
		wg.Add(1)
		go matching_numbers_between(i, i+increment, exponent, wg)
	}
	wg.Wait()
}

func main() {

	expPtr := flag.Int("power", 5, "exponent to check sums for")
	flag.Parse()
	fmt.Printf("%d: ", *expPtr)

	for i := 0; i < 10; i++ {
		cache[i] = int64(math.Pow(float64(i), float64(*expPtr)))
	}

	matching_numbers(float64(*expPtr))
	fmt.Println("done")
}
